File: src/renderer/gpu/frame.ts
import { GPUContext } from "./device";
import { DigitalRenderer, SignalPipeline } from "./digital";
import { LineBatch } from "./lines";
import { RectBatch } from "./rect";
import { TextBatch } from "./text";
import { LabelBatch } from "./labels";
import { Viewport } from "./data";
import { GpuTimer } from "./timing";

// One pill's slice of the shared pill rect/text buffers. Per-pill *draws* (not
// buffers) preserve the painter's-order occlusion — each pill's rect draws
// before, and so under, the next pill's rect, covering earlier pills' text.
export interface PillRange {
  rectStart: number; rectCount: number;
  textStart: number; textCount: number;
}

export interface FrameLayers {
  linesBg: LineBatch;
  rectsBg: RectBatch;
  // Multi-bit value labels — instanced, GPU-positioned + culled. Drawn on top of
  // the digital pipelines (inside the pills).
  labels: LabelBatch;
  // Single-pipeline value labels — the boolean format's true/false text, drawn
  // over the high/low line (same renderer, separate instance buffer + segments).
  labelsSingle: LabelBatch;
  linesFg: LineBatch;
  textBody: TextBatch;
  // All pills (markers + cursor) share one rect buffer + one text buffer (filled
  // in one writeBuffer each per frame); pillRanges[i] addresses pill i's slice,
  // drawn individually via firstInstance to keep the per-pill occlusion order.
  pillRects: RectBatch;
  pillText: TextBatch;
  pillRanges: PillRange[];
  pillRangeCount: number;
}

function drawLines(pass: GPURenderPassEncoder, b: LineBatch): void {
  if (b.lineCount === 0) return;
  pass.setPipeline(b.pipeline);
  pass.setBindGroup(0, b.bindGroup);
  pass.draw(4, b.lineCount);
}
function drawRects(pass: GPURenderPassEncoder, b: RectBatch): void {
  if (b.rectCount === 0) return;
  pass.setPipeline(b.pipeline);
  pass.setBindGroup(0, b.bindGroup);
  pass.draw(4, b.rectCount);
}
function drawText(pass: GPURenderPassEncoder, b: TextBatch): void {
  if (b.glyphCount === 0) return;
  pass.setPipeline(b.pipeline);
  pass.setBindGroup(0, b.bindGroup);
  pass.draw(4, b.glyphCount);
}
function drawLabels(pass: GPURenderPassEncoder, b: LabelBatch): void {
  if (b.glyphCount === 0 || !b.bindGroup) return;
  pass.setPipeline(b.pipeline);
  pass.setBindGroup(0, b.bindGroup);
  pass.draw(4, b.glyphCount);
}

const CLEAR_VALUE = { r: 0.106, g: 0.114, b: 0.129, a: 1 };

export function renderFrame(
  { device, ctx }: GPUContext,
  renderer: DigitalRenderer,
  pipelines: SignalPipeline[],
  layers: FrameLayers,
  vp: Viewport,
  gpuTimer?: GpuTimer,
): void {
  renderer.writeViewport(vp);

  const enc = device.createCommandEncoder();
  const pass = enc.beginRenderPass({
    colorAttachments: [{
      view: ctx.getCurrentTexture().createView(),
      loadOp: "clear",
      clearValue: CLEAR_VALUE,
      storeOp: "store",
    }],
    timestampWrites: gpuTimer?.begin(),
  });

  drawLines(pass, layers.linesBg);
  drawRects(pass, layers.rectsBg);

  for (const pipeline of pipelines) {
    pass.setPipeline(pipeline.pipeline);
    pass.setBindGroup(0, pipeline.bindGroup);
    pass.draw(4, pipeline.segmentCount);
  }

  // Value labels sit inside the multi-bit pills → draw after the digital pass.
  drawLabels(pass, layers.labels);
  // Boolean true/false labels over the single-bit lines.
  drawLabels(pass, layers.labelsSingle);

  drawText(pass, layers.textBody);
  drawLines(pass, layers.linesFg);

  // Pill overlays draw last — opaque, on top of everything else. One draw per
  // pill (rect then text) into the shared buffers via firstInstance, so each
  // pill's rect fully occludes earlier pills on overlap.
  const { pillRects, pillText, pillRanges } = layers;
  for (let i = 0; i < layers.pillRangeCount; i++) {
    const pr = pillRanges[i];
    if (pr.rectCount > 0) {
      pass.setPipeline(pillRects.pipeline);
      pass.setBindGroup(0, pillRects.bindGroup);
      pass.draw(4, pr.rectCount, 0, pr.rectStart);
    }
    if (pr.textCount > 0) {
      pass.setPipeline(pillText.pipeline);
      pass.setBindGroup(0, pillText.bindGroup);
      pass.draw(4, pr.textCount, 0, pr.textStart);
    }
  }

  pass.end();
  gpuTimer?.resolve(enc);

  device.queue.submit([enc.finish()]);
  gpuTimer?.readback();
}

File: crates/riptide-render/src/frame.rs
//! The frame encoder (port of `gpu/frame.ts`): one render pass with clear,
//! painter's order — linesBg → rectsBg → digital pipelines → labels →
//! labelsSingle → textBody → linesFg → per-pill rect+text overlays (each
//! pill's own ranged draws so its rect occludes earlier pills).
//!
//! OWNED BY UNIT U12. Calls only the frozen batch APIs of U6/U7/U8.

use riptide_contract::geometry::PillRange;

use crate::device::Gpu;
use crate::digital::SignalPipeline;
use crate::labels::LabelBatch;
use crate::lines::LineBatch;
use crate::rect::RectBatch;
use crate::text::TextBatch;
use crate::timing::GpuTimer;

/// Background clear, matches `frame.ts CLEAR_VALUE`.
pub const CLEAR_VALUE: wgpu::Color = wgpu::Color { r: 0.106, g: 0.114, b: 0.129, a: 1.0 };

pub struct FrameLayers<'a> {
    pub lines_bg: &'a LineBatch,
    pub rects_bg: &'a RectBatch,
    /// Single + multi digital pipelines (0..2 live ones).
    pub digital: &'a [&'a SignalPipeline],
    pub labels: Option<&'a LabelBatch>,
    pub labels_single: Option<&'a LabelBatch>,
    pub text_body: &'a TextBatch,
    pub lines_fg: &'a LineBatch,
    pub pill_rects: &'a RectBatch,
    pub pill_text: &'a TextBatch,
    pub pill_ranges: &'a [PillRange],
}

// The drawX helpers of frame.ts: skip empty batches.
fn draw_lines(pass: &mut wgpu::RenderPass<'_>, b: &LineBatch) {
    if b.line_count > 0 {
        b.draw(pass);
    }
}
fn draw_rects(pass: &mut wgpu::RenderPass<'_>, b: &RectBatch) {
    if b.rect_count > 0 {
        b.draw(pass);
    }
}
fn draw_text(pass: &mut wgpu::RenderPass<'_>, b: &TextBatch) {
    if b.glyph_count > 0 {
        b.draw(pass);
    }
}
fn draw_labels(pass: &mut wgpu::RenderPass<'_>, b: Option<&LabelBatch>) {
    if let Some(b) = b
        && b.glyph_count > 0
    {
        b.draw(pass);
    }
}

/// Encodes + submits one frame into `view`.
pub fn render_frame(
    gpu: &Gpu,
    view: &wgpu::TextureView,
    layers: &FrameLayers<'_>,
    mut timer: Option<&mut GpuTimer>,
) {
    let mut enc = gpu
        .device
        .create_command_encoder(&wgpu::CommandEncoderDescriptor { label: Some("frame") });
    {
        let mut pass = enc.begin_render_pass(&wgpu::RenderPassDescriptor {
            label: Some("frame-pass"),
            color_attachments: &[Some(wgpu::RenderPassColorAttachment {
                view,
                depth_slice: None,
                resolve_target: None,
                ops: wgpu::Operations {
                    load: wgpu::LoadOp::Clear(CLEAR_VALUE),
                    store: wgpu::StoreOp::Store,
                },
            })],
            depth_stencil_attachment: None,
            timestamp_writes: timer.as_deref().and_then(GpuTimer::pass_timestamp_writes),
            occlusion_query_set: None,
        });

        draw_lines(&mut pass, layers.lines_bg);
        draw_rects(&mut pass, layers.rects_bg);

        for pipeline in layers.digital {
            pipeline.draw(&mut pass);
        }

        // Value labels sit inside the multi-bit pills → draw after the digital
        // pipelines; then the boolean true/false labels over the 1-bit lines.
        draw_labels(&mut pass, layers.labels);
        draw_labels(&mut pass, layers.labels_single);

        draw_text(&mut pass, layers.text_body);
        draw_lines(&mut pass, layers.lines_fg);

        // Pill overlays draw last — opaque, on top of everything else. One
        // ranged draw per pill (rect then text) into the shared buffers, so
        // each pill's rect fully occludes earlier pills on overlap.
        for pr in layers.pill_ranges {
            if pr.rect_count > 0 {
                layers.pill_rects.draw_range(&mut pass, pr.rect_start, pr.rect_count);
            }
            if pr.text_count > 0 {
                layers.pill_text.draw_range(&mut pass, pr.text_start, pr.text_count);
            }
        }
    }

    if let Some(t) = timer.as_deref_mut() {
        t.resolve(&mut enc);
    }
    gpu.queue.submit([enc.finish()]);
    if let Some(t) = timer {
        t.readback();
    }
}
